### PR TITLE
[JENKINS-29956] Test Windows junctions before Java 7 symlink

### PR DESCRIPTION
**Pull request by dbroady1**
Apparently Java 7 (and later) do not count Windows junctions as
symlinks.  When you drill down into the BasicFileAttributes structure,
reparse points are only counted under isOther.  So, since we already
have code that properly detects Windows Junctions, let's use that first
and then fallback to the Java 7 code.

### CODE REVIEW

**Comment by jtnord** [comment]
@dbroady1 can you elaborate on the issue this is fixing - rather than the issue in the function that it is is addressing.
Linux doesn't support symlinks on directories - and the equivalent of junctions would be bind mounts - which are not handled specially AFAICT.  Also whilst Jenkins will create symlinks - it does not to my knowledge create Junctions.

**Comment by dbroady1** [comment]
Sure.

On my Windows box, Jenkins is calling out to my custom build script, and part of that build script is creating a junction in the workspace folder to a different location (a machine cache of code, so that it's not duplicated multiple times on the machine).  On Linux, the junction is replaced with a symlink to a directory.

When my project is built, I ask for the workspace to be wiped.  It does that, but on Windows only, it's proceeding through the junction and deleting the pointed to data, instead of just removing the junction.  On Linux, it just removes the symlink itself & does not recurse.

It is my understanding that making FilePath think that the junction is a symlink will prevent it from recursing down into junction and deleting those contents.  Am I correct on that assumption?

As for Linux doesn't support symlinks on directories, I guess we have a terminology problem here.  In Linux, I can easily do the following:
   %  cd /tmp
   %  mkdir darin && cd darin
   %  ln -s /var .
   %  ls -l
lrwxrwxrwx 1 darin darin 4 Aug  6 16:49 var -> /var
That appears to be a directory symlink for me, which is what I see Windows junctions doing.  For instance, on my Windows box, if I do:
    % cd c:\temp
    % mklink /J W2K c:\Windows
    % dir
08/06/2015 04:54 PM  <JUNCTION>  W2K [c:\Windows].

Linux symlinks can span file systems, as can Windows junctions.  However, Linux hardlinks & Windows hardlinks & symlinks can not.

**Comment by jtnord** [comment]
> Linux doesn't support symlinks on directories

it's late and I had a brain fart.  as you point out its hardlinks that aren't supported...

**Comment by jtnord** [comment]
@dbroady1 thanks for the explanation.

> Windows hardlinks & symlinks can not [span file systems]

You can, unless I am very much mistaken, create symlinks that span volumes **and even computers** on any MS platform currently in mainstream support.
Windows 7 even allows you to create a <symlinkd> to a remote share with `mklink /d music \\nas\music` that cmd.exe is perfectly happy navigating around -  this works just as well for a local volume as well.

I don't think Jenkins would be well behaved if you used a Linux bind mount instead of a symlink either (which is probaly the equivallent of a windows junction)

Irrespective of all the above Jenkins may soon be migrating to using java7 APIs directly rather than via reflection so this code would be likely to be removed in a future version unless it was commented with a reason not to do so!  
So I would suggest:
* create a jenkins JIRA that can be referenced in the commit message
* add a comment to say don't change the order delete this code because xxx with a link to the JIRA so people can get more information so that it is not reverted in the future.
* A unit test as well would prevent any accidents and surprises if you upgraded in the future.

**Comment by dbroady1** [comment]
I've created issue #29956 to reflect this.  I will update the code with a comment stating why this is necessary.

**Comment by jtnord** [comment]
[JENKINS-29956](https://issues.jenkins-ci.org/browse/JENKINS-29956)

**Comment by jtnord** [comment]
@dbroady1 could you write a unit test to verify this on windows?

otherwise code changes look good to me.

**Comment by daniel-beck** [comment]
Ideally the commit message references the Jira issue as well.

File: core/src/main/java/hudson/Util.java [hunk 1]
```diff
@@ -331,10 +331,20 @@ public static void deleteRecursive(@Nonnull File dir) throws IOException {
...
+         *  It thinks of them as normal directories.  To use the NIO2 API & treat it like a symlink,
+         *  you have to go through BasicFileAttributes and do the following check:
+         *     isSymbolicLink() || isOther()
+         *  The isOther() call will include Windows reparse points, of which a directory junction is.
+         *
+         *  Since we already have a function that detects Windows junctions or symlinks and treats them
+         *  both as symlinks, let's use that function and always call it before calling down to the
+         *  NIO2 API.
+         *
+         *  This is tracked as JENKINS-29956.
```
Comment (daniel-beck) [inline core/src/main/java/hudson/Util.java]: Assuming this change resolves JENKINS-29956, the issue should not be referenced in the comment. This could be mistaken as a known issue or TODO otherwise.

Comment (jtnord) [inline core/src/main/java/hudson/Util.java]: @daniel-beck I respectfully disagree.  The reason the code is like this is due to JENKINS-29956.  The comment should direct devs to a place where they can find out more information and `git blame` via the github UI does not support ignoring whitespace changes ad as such can not be relied upon to link these changes to the actual reason for the change in the future.

Now if you want to request the link is handled differently - "for the rationale here see JENKINS-XXXX" but it is important to me that this link is left in the code.

Comment (daniel-beck) [inline core/src/main/java/hudson/Util.java]: The rationale for this change is that the previous code was obviously wrong, as explained quite nicely by the comment. It considers junctions to be like symlinks in a few lines below, but having the Java 7 check (that only considers real symlinks) early breaks this behavior. While it resulted in JENKINS-29956, that's not required to understand why this change was performed.

**Comment by oleg-nenashev** [comment]
:+1:

**Comment by jtnord** [comment]
@dbroady1 can you write a unit test for this to ensure the behavior does not regress.

**Comment by oleg-nenashev** [comment]
@jtnord @dbroady1 
From what I see this PR is still actual

**Comment by dbroady1** [comment]
I added the unit test requested by jtnord.  It will only run on Windows.  I ran the unit test prior to my patch, and it failed.  With my patch, it works.

**Comment by batmat** [comment]
Is that PR as-is still useful with Jenkins now requiring Java 7 or above to run? 
Also, I thought I had inline `isSymlinkJava7()` in #2256 seems like I missed it...

**Comment by dbroady1** [comment]
Yes, it is still valid.  The reason is that java.nio.file.Path.isSymbolicLink is the culprit.  It does not think of DirectoryJunctions as symbolic links.

File: core/src/main/java/hudson/Util.java (hunk 1 above)
Comment (jtnord) [inline core/src/main/java/hudson/Util.java]: @daniel-beck so are you just saying that line 346 should be removed, or complaining about the whole comment.  I think I miss understood the first time around.

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        print(j1);
+        Util.deleteRecursive(j1);
+        print(dir1Root);
+        print(tmp.getRoot());
+        assertTrue(f1.exists());
+    }
+
+    private File makeJunction(File baseDir, File pointToDir) throws Exception {
+       File junc = new File(baseDir, "testJunction");
+       Process p = Runtime.getRuntime().exec("cmd.exe /C \"mklink /J " + junc.getPath() + " " + pointToDir.getPath() + "\"");
```
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: Will this test not likely break if you run it in a folder with spaces.. (c:\ Documents and Settings\Foobar\Temp)?

also mklink is an executable, so just call it correctly without wrapping it inside a cmd.exe.

It would be better to use the Array form of [`Runtime.exec`](https://docs.oracle.com/javase/7/docs/api/java/lang/Runtime.html#exec(java.lang.String[])) here

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+       p.waitFor();
```
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: check the return code is zero  - and skip the test if not (`Asume`...)

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+
+    @Before
+    public void windowsOnly() {
+       assumeTrue(File.pathSeparatorChar==';');
+    }
+
+    @Test
+    @Issue("JENKINS-2995")
+    public void testJunctionIsRemovedButNotContents() throws Exception {
+        TemporaryFolder dir1 = new TemporaryFolder(tmp.getRoot());
```
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: unneeded.

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        dir1.create();
```
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: [javadoc](http://junit.org/junit4/javadoc/latest/org/junit/rules/TemporaryFolder.html#create()) says do not use?

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+        File dir1Root = dir1.getRoot();
+        File f1 = dir1.newFile();
+        File j1 = makeJunction(tmp.getRoot(), dir1Root);
```
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: create directories under tmp instead to make sure cleanup happens correctly.  (`tmp.newFolder(...)`)

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import org.junit.rules.TemporaryFolder;
+import org.jvnet.hudson.test.Issue;
+
+public class RemoveWindowsDirectoryJunctionTest {
+    @Rule
+    public TemporaryFolder tmp = new TemporaryFolder();
+
+    @Before
+    public void windowsOnly() {
+       assumeTrue(File.pathSeparatorChar==';');
```
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: NIT: can use Functions.isWindows()

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+        File dir1Root = dir1.getRoot();
+        File f1 = dir1.newFile();
+        File j1 = makeJunction(tmp.getRoot(), dir1Root);
+        print(tmp.getRoot());
+        print(dir1Root);
+        print(j1);
+        Util.deleteRecursive(j1);
+        print(dir1Root);
+        print(tmp.getRoot());
+        assertTrue(f1.exists());
```
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: Generally put a message with your assertions, so when the tests fail on the CI it is more obvious why without having to look at the code.
`assertTrue("Contents of the junction should not be deleted", f1.exists())`

**Comment by jtnord** [comment]
I think there are some issues with the unit test.

**Comment by dbroady1** [comment]
I will fix all the issues listed.

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java (hunk 2 above)
Comment (dbroady1) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: According to http://ss64.com/nt/syntax-internal.html, mklink is an internal cmd.exe command, not an external exe.  So, my use of cmd.exe /c "mklink /J ..." is correct.  I will fix it so that it will work with spaces in the path name, however.

File: core/src/main/java/hudson/Util.java (hunk 1 above)
Comment (daniel-beck) [inline core/src/main/java/hudson/Util.java]: @jtnord I'm specifically addressing line 346.

Comment (dbroady1) [inline core/src/main/java/hudson/Util.java]: Just to be specific -- you want me to remove the line "This is tracked as JENKINS-29956", correct?

Comment (jtnord) [inline core/src/main/java/hudson/Util.java]: @dbroady1 yes - that is what @daniel-beck is asking for.

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java (hunk 2 above)
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: yes - my mistake...

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        assertTrue("Contents of Windows Junction should not be removed", f1.exists());
+    }
+
+    private File makeJunction(File baseDir, File pointToDir) throws Exception {
+       File junc = new File(baseDir, "test Junction");
+       String cmd = "mklink /J \"" + junc.getPath() + "\" \"" + pointToDir.getPath() + "\"";
+       ProcessBuilder pb = new ProcessBuilder("cmd.exe", "/C", cmd);
+       pb.inheritIO();
+       Process p = pb.start();
+       assertTrue("Running mklink failed (cmd=" + cmd + ")", p.waitFor() == 0);
```
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: NIT, the error code may be usefull.  `assertEquals(""Running mklink failed (cmd=" + cmd + ")", 0, p.waitFor());`

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+        File j1 = makeJunction(tmp.getRoot(), subdir1);
+        System.out.println("Before Util.deleteRecursive(j1)");
+        print(tmp.getRoot());
+        print(subdir1);
+        print(j1);
+        Util.deleteRecursive(j1);
+        System.out.println("After Util.deleteRecursive(j1)");
+        print(subdir1);
+        print(tmp.getRoot());
+        assertTrue("Contents of Windows Junction should not be removed", f1.exists());
```
Comment (jtnord) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: NIT: possibly also `assertFalse("junction should have been removed", j1.exists())`?

**Comment by jtnord** [comment]
LGTM baring the comment change in the code that has been pointed out.
A couple of NITs in the unit test, if you feel like addressing them when you are changing the (c), but not a deal breaker if you don't

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+        File subdir1 = tmp.newFolder("notJunction");
+        File f1 = new File(subdir1, "testfile1.txt");
+        assertTrue("Unable to create temporary file in notJunction directory", f1.createNewFile());
+        File j1 = makeJunction(tmp.getRoot(), subdir1);
+        System.out.println("Before Util.deleteRecursive(j1)");
+        print(tmp.getRoot());
+        print(subdir1);
+        print(j1);
+        Util.deleteRecursive(j1);
+        System.out.println("After Util.deleteRecursive(j1)");
```
Comment (batmat) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: Do not put sysouts in tests, IMO.

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Test
+    @Issue("JENKINS-2995")
+    public void testJunctionIsRemovedButNotContents() throws Exception {
+        File subdir1 = tmp.newFolder("notJunction");
+        File f1 = new File(subdir1, "testfile1.txt");
+        assertTrue("Unable to create temporary file in notJunction directory", f1.createNewFile());
+        File j1 = makeJunction(tmp.getRoot(), subdir1);
+        System.out.println("Before Util.deleteRecursive(j1)");
```
Comment (batmat) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: Do not put sysouts in tests, IMO.

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        print(tmp.getRoot());
+        print(subdir1);
+        print(j1);
```
Comment (batmat) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: IMO, this shouldn't be here. I'd remove those prints and the associated utility function.

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+    }
+    
+    private void print(File d) {
+        System.out.println("  '" + d.getPath() + "'");
+        String[] c = d.list();
+        for (String s : c) {
+            System.out.println("    '" + s + "'");
+        }
+    }
+}
```
Comment (batmat) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: NIT: missing newline.

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java (hunk 9 above)
Comment (batmat) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: NIT: might be a bit better to do `assertEquals("Running mklink failed (cmd=" + cmd + ")", 0, p.waitFor());` so that JUnit prints out the received value in case of failure.

File: core/src/main/java/hudson/Util.java (hunk 1 above)
Comment (batmat) [inline core/src/main/java/hudson/Util.java]: Somewhat NIT: add \<p> for your paragraphs or it's gonna be rendered not like you seem to expect it, seeing how you formatted it (in the IDE for example), and so will be less readable.

Comment (jtnord) [inline core/src/main/java/hudson/Util.java]: @batmat its not javadoc - it is an inline comment - it does not get rendered :-)

If you IDE is reformatting comments then turn it off :-)

Comment (batmat) [inline core/src/main/java/hudson/Util.java]: @jtnord yup, read too quickly. I thought I had seen two ** :)

File: core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java (hunk 9 above)
Comment (batmat) [inline core/src/test/java/hudson/RemoveWindowsDirectoryJunctionTest.java]: Just realized I posted somewhat the same comment as @jtnord. Seems like I kept that tab open too long w/o refreshing :). Weird, it thought GitHub was generally doing live updates. Maybe not inline here.

**Comment by oleg-nenashev** [comment]
@jtnord @batmat 
Seems it needs another review

**Comment by oleg-nenashev** [comment]
@jtnord @batmat ping

**Comment by oleg-nenashev** [comment]
Retriggering the full build. Windows test still passes for me

**Comment by jtnord** [comment]
:+1:

**Comment by oleg-nenashev** [comment]
👍 from me as well
